core/new-gui/src/app/workspace/service/workflow-graph/model/coeditor-presence.service.ts

```diff
@@ -0,0 +1,316 @@
+import { Injectable } from "@angular/core";
+import { WorkflowGraph } from "./workflow-graph";
+import * as joint from "jointjs";
+import { JointGraphWrapper } from "./joint-graph-wrapper";
+import { User, UserState } from "../../../../common/type/user";
+import { WorkflowActionService } from "./workflow-action.service";
+import { JointUIService } from "../../joint-ui/joint-ui.service";
+import { Observable, Subject } from "rxjs";
+
+function isEqual(array1: any[] | undefined, array2: any[] | undefined): boolean {
+  if (!array1 && !array2) return true;
+  else {
+    if (array1 && array2) {
+      return array1.length === array2.length && array1.every((value, index) => value === array2[index]);
+    } else return false;
+  }
+}
+
+/**
+ * CoeditorPresenceService handles user-presence updates from other editors in the same shared-editing room
+ * and shows them on the UI. It also keeps some states in itself for some necessary UI update information,
+ * like which co-editors are currently highlighting a particular operator.
+ */
+
+@Injectable({
+  providedIn: "root",
+})
+export class CoeditorPresenceService {
+  private readonly coeditorOpenedCodeEditorSubject = new Subject<{ operatorId: string }>();
+  private readonly coeditorClosedCodeEditorSubject = new Subject<{ operatorId: string }>();
+  public shadowingModeEnabled = false;
+  public shadowingCoeditor?: User;
+  public coeditors: User[] = [];
+  private jointGraph: joint.dia.Graph;
+  private texeraGraph: WorkflowGraph;
+  private jointGraphWrapper: JointGraphWrapper;
+  private coeditorCurrentlyEditing = new Map<string, string | undefined>();
+  private coeditorOperatorHighlights = new Map<string, string[]>();
+  private coeditorOperatorPropertyChanged = new Map<string, string | undefined>();
+  private coeditorEditingCode = new Map<string, boolean>();
+  private coeditorStates = new Map<string, UserState>();
+  private currentlyEditingTimers = new Map<string, NodeJS.Timer>();
+
+  constructor(private workflowActionService: WorkflowActionService) {
+    this.texeraGraph = workflowActionService.getTexeraGraph() as WorkflowGraph;
+    this.jointGraph = workflowActionService.getJointGraph();
+    this.jointGraphWrapper = workflowActionService.getJointGraphWrapper();
+    this.observeUserState();
+    this.texeraGraph.newYDocLoadedSubject.subscribe(_ => {
+      this.observeUserState();
+    });
+  }
+
+  /**
+   * Listens to changes of co-editors' presence infos and lets <code>{@link CoeditorPresenceService}</code> handle them.
+   */
+  private observeUserState(): void {
+    // destroy previous user states if any
+    for (const coeditor of this.coeditors) {
+      if (coeditor.clientId) this.removeCoeditor(coeditor.clientId);
+    }
+
+    // first time logic
+    const currentStates = Array.from(
+      this.texeraGraph.sharedModel.awareness.getStates().values() as IterableIterator<UserState>
```

Comment:
we can wrap `this.texeraGraph.sharedModel.awareness.getStates()` into another function and annotate proper return types, this could reduce many type casts.

Reply:
Done, also refactored some code with similar problems.